### DISCUSSION TITLE
Is there a way to automatically provide types for a "higher order" createMachine?

### DISCUSSION THREAD

**Discussion by nstadigs** [statelyai/xstate#4315] [General] [answered]
In my system events, actions, guards etc are provided by the framework and are the same for every machine. Is there an easy way to automatically provide these types for the machine configs?

Something like (pseudoish code):
```
type SystemSpecificMachineConfig = // Extends default xstates internal config type somehow.

function createSystemSpecificMachine<SystemSpecificMachineConfig>(config) {
  return createMachine(config);
}

const myActor = createSystemSpecificMachine({
  // No need to specify types here. Event-, action-, some guard types etc for this config are provided by the framework.
  id: "dishwasher state",
  initial: "idle",
  states: {
    idle: {
      on: {
        "sensor.dishwasher.energy": { // Valid event provided by the system
          target: "running",
          guard: ({ event }) => {
            return event.value > 30 // event.value is number
          } 
        }
      }
    },
    running: {
      on: {
        "sensor.dishwasher.energy": { // Valid event provided by the system
          target: "idle",
          guard: ({ event }) => {
            return event.value < 10 // event.value is number
          } 
        }
      }
    }
  }
});
```

... or is this the wrong way of thinking about this? Is there a way to provide these types on a system level? I don't even know how that would work tho...

Thanks!

**Comment by Andarist** [comment]
> type SystemSpecificMachineConfig = // Extends default xstates internal config type somehow.

It ain't pretty and we are not married to the exact shape/list of our current type parameters but it's possible today: [TS playground](https://www.typescriptlang.org/play?ts=5.2.2#code/JYWwDg9gTgLgBAbwLACg5wLIEMDGALYAOwFMBhCQmYgDxgBpV0BRAN2MoHkAjAK2J3qM4AQUIBPVuxjc+AhmjgAFKBBbAAJsXXCB0eekVYoWEMSpRgALy0z+ghQDkKAJS0BXHMC4AbYgFVCAGtCCAB3Qn04ABUxMGIAc3ZyQgBnGGMie3RsfCIyCgAzYHjIgElKYihCLG8cghJS8F9TSiwYYAoUyNcUiG82GLjEwgwzLEiAZRg24jq8yOUIMEicKGIZuZJ5AF84ApUQOAAiajSZo4BuOAB6a7gYWOIUgC44AFYAOgAGb4BaLjGHwAzF9UKgHnE4AB1YAwPAYCRsSgpAA8UUklDgNCohHUKTgGOkvDsAD44ABeIQAH2ihOpiCE6HQEOIryOIDEAH0cG40hAQJziEiYJdGXBthcwSgaJBYHs3IQBB1CHBVusqAjNsQUUIoskqLQsbR2HjMLh6vkKrRIujhUacabRIipLY5LqdDBoPaTfjFmpNNpdFAbR7ld7cb6jCYzJUrDZiW6FFEAOJuIzqcOmwzGUzmOPqV1ZaIAEWI3iwYkz+LSFkIJV1USw8SrcBrRHrSfKYDcRaiHB73fg2J9cCchFc6g8Xl8AWCYQiDceKVG0xbgwSSU66SwmXkJIAFEIcIViq8tckivEdQp0HqKAai+gYXCEYTUbapCTIrePXoxVFQwob9olTdNgKiUtyzEcDGw7Jloi7HtwP7GBBxgpcVywIQvyEUAwGaKQ2mVFIAH5XnKcxqlqc08kafDiBaaZ2k6a94LvK1H2hWF4WdZE0UJHCb2iX9g3-QCFyEiCywrYCej6AZHmGTDWPg6J12XMZgKfbjX2Fd8BK04Sg0MgClSAsVb1AqB1BMyCZIstSmzFElsPkABKV4pg2GiSBU9iH0iZ8eLffjhUEn9jPdMyJMstNrJtOzoKERYwBU9A5P6Yh1yUsY0qZKJ1Mwwygt0qR9LCkyRMq6KTKsmyHKkqCTNg5yAG0jjWXpMvUI4AF1gKOaZ4hSI5sLgAAyMU2zrG1EN7FC0KEDKFKGdhlP-QrNLFEreJgcrP3AqqxJq-86vAxKYKchQXJQMlkAUNYYDcKpVTWbzchIfdj0IS86DgPCCNaZjUjcyUUG2KU1Q2MQtX3e7mSXV4EF2LB8XhpkhTKpGHJZNkADEOA4UUhIlIRtkiChsaE257igSssB7CBjyaGM4DwSpiDJnZQdQIA)

**Reply by nstadigs** [reply]
Thanks a bunch! Seems like the type of events also must be specified by the "types" property for this to work, no? If I remove that the autocomplete for event names stops working.

**Reply by Andarist** [reply]
Could you showcase the exact problem? I see the completions here using this unmodified playground:
<img width="540" alt="Screenshot 2023-09-27 at 15 54 11" src="https://github.com/statelyai/xstate/assets/9800850/9a56b7ef-d376-45c9-a0a9-76cd4c2891ce" alt="completions displayed from both sources">

**Reply by nstadigs** [reply]
I mean, I expected to see only "my_custom_event" after removing the "types" property from your example, but instead I get nothing.

**Reply by Andarist** [reply, marked as answer]
You can fix this by pushing `TEvent` in your custom `createMyMachine` to the end and by setting its default to `never`: [TS playground](https://www.typescriptlang.org/play?ts=5.2.2#code/JYWwDg9gTgLgBAbwLACg5wLIEMDGALYAOwFMBhCQmYgDxgBpV0BRAN2MoHkAjAK2J3qM4AQUIBPVuxjc+AhmjgAFKBBbAAJsXXCB0eekVYoWEMSpRgALy0z+ghQDkKAJS0BXHMC4AbYgFVCAGtCCAB3Qn04ABUxMGIAc3ZyQgBnGGMie3RsfCIyCgAzYHjIgElKYihCLG8cghJS8F9TSiwYYAoUyNcUiG82GLjEwgwzLEiAZRg24jq8yOUIMEicKGIZuZJ5AF84ApUQOAAiajSZo4BuOAB6a7gYWOIUgC44AFYAOgAGb4BaLjGHwAzF9UKgHnE4AB1YAwPAYCRsSgpAA8UUklDgNCohHUKTgGOkvDsAD44ABeIQAH2ihOpiCE6HQEOIryOIDEAH0cG40hAQJziEiYJdGXBthcwSgaJBYHs3IQBB1CHBVusqAjNsQUUIoskqLQsbR2HjMLh6vkKrRIlEdDBoEacabFmpNNpdFAbXblY6TfjDMZTOYrDZiXJdQBxNxGdS+3H+owmMyVEPqWzhhRRAAixG8WDEcdNaQshBKuqiWHihfxxaIZcz5TAbiy0Q4zab8GxfrgTkIrnUHi8vgCwTCEXLjxSo2m1eij2GyWLWEyNsJs9EiKk6fg5LgJDYntQJIAFEIcIViq8tckivEdQp0HqKAaW+gYXCEYTUejhSTIo+7T0MVbSVCh-2iKMY3A7Nc3zaCK3rJlokbZtoLbGAO3gydpywIQ-yEUAwGaKQ2mVFIAH5XnKcxqlqc08kaIjiBaaZ2k6e8kKfK1X2hWF4U3ZE0UJfCH2iQDD1EkC2PHSSczzMRwJ6PoBnndgcI4pDokGJ4cPAt8+M-YVv2EvSxI9UypOVCzIKgdQLLkuCxUfBCxRJPD5AASleKYNnokgNK4l9InffivyE38vXM3VvTAyNo1sm0HIUoRFjADT0CU-piG04Z1KcrTsLGUyQsMqRjIi-KpKA0SANAmTNIg+K7MqpKLJc0SSQAbSONZeiy9QjgAXXAo5pniFIjjwuAADIxVrUsbRQlsonQzChEylShjUsZ0oKuIpyKsUSoEmByqkETOPE6CYvqpkohs5rZNg5LJPa9A3JQMlkAUNYYDcKpVTWXzchIY9z0IW86DgQjiNaaSUg8yUUG2KU1Q2MQtWPb7mUnV4EF2LB8WxplbixIy8bFUmcbiNkADEOA4UVRNJiUhG2SIKAp5m7nSAssGbCBzyaZM4DwSpiDFNmdkR1AgA)

**Reply by nstadigs** [reply]
Wonderful! You really are a wizard

I guess I should pin the xstate version now tho ;)

**Reply by Andarist** [reply]
> I guess I should pin the xstate version now tho ;)

Yes, this is very much advised now. As much as I'd like to give you guarantees related to those type parameters the truth is that we are pushing TS to its limits, trying a lot of things to improve the final DX. By limiting our right to change/tweak those we'd have to forego any kind of experimentation that could benefit most of the XState users.

**Reply by nstadigs** [reply]
I won't hold you to it! Anyhow, being able to do this seems like a nice feature and opens up the possibility of building more targeted libraries/frameworks on top of xstate.
